src/BrainPass/BrainPass.sol

```diff
@@ -0,0 +1,402 @@
+
+        emit TimeIncreased(
+            msg.sender,
+            tokenId,
+            pass.startTimestamp,
+            pass.endTimestamp
+        );
+    }
+
+    /// @notice Withdraws any amount in the contract
+    function withdrawEther() external payable onlyOwner {
+        uint256 ethbalance = address(this).balance;
+        if (ethbalance <= 0) revert NoEtherLeftToWithdraw();
+
+        (bool success, ) = (msg.sender).call{value: ethbalance}("");
+        if (!success) revert TransferFailed();
+    }
+
+    function withdrawIQ() external payable onlyOwner {
+        uint256 tokenBalance = IERC20(iqToken).balanceOf(address(this));
+
+        if (tokenBalance <= 0) revert NoIQLeftToWithdraw();
+
+        bool tokenSuccess = IERC20(iqToken).transfer(msg.sender, tokenBalance);
+        if (!tokenSuccess) revert TransferFailed();
+    }
+
+    /// -----------------------------------------------------------------------
+    /// Internal Functions
+    /// -----------------------------------------------------------------------
+
+    /// @notice Calculate the price of an Nft
+    /// @param startTimestamp The start time to calculate the price of the Nft
+    /// @param endTimestamp The end time to calculate the price of the Nft
+    function calculatePrice(
+        uint256 passId,
+        uint256 startTimestamp,
+        uint256 endTimestamp
+    ) internal view returns (uint256) {
+        PassType memory passType = passTypes[passId];
+        uint256 subscriptionPeriodInSeconds = endTimestamp - startTimestamp;
+
+        uint256 subscriptionPeriodInDays = subscriptionPeriodInSeconds / 1 days;
+
+        // Calculate the total price
+        uint256 totalPrice = subscriptionPeriodInDays * passType.pricePerDay;
+
+        if (passType.discount > 0) {
+            uint256 discountAmount = (totalPrice * passType.discount) / (100);
+            totalPrice = totalPrice - discountAmount;
+        }
+
+        return totalPrice;
+    }
+
+    /// @notice Validates the Timestamp Duration for minting Nft
+    /// @param startTimestamp The start time for checking the validity of a pass
+    /// @param endTimestamp The end time for checking the validity of a pass
+    function validatePassDuration(
+        uint256 startTimestamp,
+        uint256 endTimestamp
+    ) internal view returns (bool) {
+        uint256 durationInDays = (endTimestamp - startTimestamp) / 1 days;
+        if (endTimestamp < block.timestamp) {
+            revert DurationNotInTimeFrame();
+        }
+        return
+            durationInDays >= MINT_LOWER_LIMIT &&
+            durationInDays <= MINT_UPPER_LIMIT;
+    }
+
+    /// -----------------------------------------------------------------------
+    /// Getters
+    /// -----------------------------------------------------------------------
+
+    /// @notice Gets the tokenUri for the contract
+    function _baseURI() internal view virtual override returns (string memory) {
```

Comment:
you prob. dont need this if you use erc721

Reply:
Im overriding it here cos in erc721 it returns an empty string, hence overiding